File: scripts/shuffle_answers.js
/**
 * 打乱题库选项顺序脚本 v2（纯字符串处理，不走 JSON）
 * - 打乱 A/B/C/D 选项顺序
 * - 更新 answer 索引
 * - 跳过包含"以上都"的题目
 * - 跳过判断题(tf)
 */
const fs = require('fs');
const path = require('path');

function shuffle(array) {
  const arr = [...array];
  for (let i = arr.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [arr[i], arr[j]] = [arr[j], arr[i]];
  }
  return arr;
}

// 从一道题的 JSON 字符串中提取信息
function parseQuestionBlock(str) {
  // 找到 opts 数组
  const optsMatch = str.match(/"opts":\s*\[([^\]]+)\]/);
  if (!optsMatch) return null;
  const optsStr = optsMatch[1];

  // 解析每个选项（处理中文引号）
  const opts = [];
  const optMatches = optsStr.matchAll(/"([^"]*)"/g);
  for (const m of optMatches) {
    opts.push(m[1]);
  }

  // 找 answer
  const answerMatch = str.match(/"answer":\s*(\d+|true|false)/);
  if (!answerMatch) return null;
  let answer = answerMatch[1];
  const isTrueFalse = answer === 'true' || answer === 'false';
  if (isTrueFalse) {
    return { type: 'tf', opts, answer: answer === 'true' };
  }
  answer = parseInt(answer);

  // 判断是否跳过
  const correctText = opts[answer];
  if (!correctText) return null;
  const skip = correctText.includes('以上都') || correctText.includes('以上各');
  if (skip) return { type: 'choice', opts, answer, skip: true };

  // 打乱
  const indexed = opts.map((text, idx) => ({ text, idx }));
  const shuffled = shuffle(indexed);
  const newAnswer = shuffled.findIndex(item => item.text === correctText);

  return { type: 'choice', opts: shuffled.map(s => s.text), answer: newAnswer, skip: false };
}

// 将处理后的题重建为 JSON 字符串（保留原始引号风格）
function rebuildQuestion(q) {
  const optsStr = q.opts.map(o => `"${o}"`).join(', ');
  const answerStr = typeof q.answer === 'boolean' ? q.answer : q.answer;
  return `{"type": "choice", "q": "${q.q || ''}", "opts": [${optsStr}], "answer": ${answerStr}, "hint": "${q.hint || ''}"}`;
}

// 处理判断题（不变）
function rebuildTfQuestion(str) {
  return str; // 原样返回
}

// 核心：用状态机遍历 JS 文件，找到每个 choice 题并处理
function processFile(filePath) {
  console.log(`\n处理文件: ${path.basename(filePath)}`);

  const content = fs.readFileSync(filePath, 'utf8');
  let result = '';
  let i = 0;
  let total = 0, shuffled = 0, skipped = 0;

  while (i < content.length) {
    // 找到下一个 {"type": "choice 开头的位置
    const choiceStart = content.indexOf('{"type": "choice"', i);
    if (choiceStart === -1) {
      result += content.substring(i);
      break;
    }

    // 找到这个对象的结束位置（匹配 {}）
    let brace = 0;
    let objStart = -1;
    let j = choiceStart;
    while (j < content.length) {
      if (content[j] === '{') {
        if (objStart === -1) objStart = j;
        brace++;
      } else if (content[j] === '}') {
        brace--;
        if (brace === 0 && objStart !== -1) {
          break;
        }
      }
      j++;
    }

    const objStr = content.substring(choiceStart, j + 1);
    const parsed = parseQuestionBlock(objStr);

    if (parsed && parsed.type === 'choice' && !parsed.skip) {
      total++;
      shuffled++;
      // 找到 q, opts, answer, hint 的位置并替换
      let newStr = objStr;

      // 替换 opts
      newStr = newStr.replace(/"opts":\s*\[([^\]]+)\]/, (m, oldOpts) => {
        return `"opts": [${parsed.opts.map(o => `"${o}"`).join(', ')}]`;
      });

      // 替换 answer
      newStr = newStr.replace(/"answer":\s*\d+/, `"answer": ${parsed.answer}`);

      result += content.substring(i, choiceStart) + newStr;
      i = j + 1;
    } else if (parsed && parsed.skip) {
      total++;
      skipped++;
      result += content.substring(i, j + 1);
      i = j + 1;
    } else {
      result += content.substring(i, choiceStart + 1);
      i = choiceStart + 1;
    }
  }

  fs.writeFileSync(filePath, result, 'utf8');
  console.log(`  ✅ 完成！共${total}选择题，打乱${shuffled}题，跳过${skipped}题`);
}

const dataDir = path.join(__dirname, '..', 'data');
const files = ['4-2.js', '5-2.js', '6-2.js'];

files.forEach(file => {
  const filePath = path.join(dataDir, file);
  if (fs.existsSync(filePath)) {
    processFile(filePath);
  }
});

console.log('\n🎉 全部处理完成！记得 git 提交推送。');

File: scripts/shuffle_answers.py
# -*- coding: utf-8 -*-
import sys
import io
sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')

"""
打乱题库选项顺序脚本
- 打乱 A/B/C/D 选项顺序
- 更新 answer 索引
- 跳过包含"以上都"的题目
- 跳过判断题(tf)
"""
import re
import random
import os

def fisher_yates_shuffle(arr):
    """原地洗牌"""
    a = list(arr)
    for i in range(len(a) - 1, 0, -1):
        j = random.randint(0, i)
        a[i], a[j] = a[j], a[i]
    return a

def process_file(filepath):
    print(f"\n处理文件: {os.path.basename(filepath)}")

    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()

    total = 0
    shuffled = 0
    skipped = 0
    result = content
    offset = 0  # 记录因替换导致的位置偏移

    # 遍历所有 choice 题
    # 匹配 {"type": "choice", ... } 对象
    pattern = re.compile(
        r'\{"type": "choice", "q": "([^"]*)", "opts": \[([^\]]+)\], "answer": (\d+), "hint": "([^"]*)"\}',
        re.DOTALL
    )

    matches = list(pattern.finditer(content))
    # 逆序处理（从后往前替换，避免索引偏移问题）
    for m in reversed(matches):
        q_text = m.group(1)
        opts_str = m.group(2)
        answer_idx = int(m.group(3))
        hint = m.group(4)

        # 解析选项数组
        opts = re.findall(r'"([^"]*)"', opts_str)

        total += 1

        # 判断是否跳过
        if answer_idx >= len(opts):
            continue
        correct_text = opts[answer_idx]
        if '以上都' in correct_text or '以上各' in correct_text:
            skipped += 1
            continue

        # 打乱
        indexed = [(t, i) for i, t in enumerate(opts)]
        shuffled_idx = fisher_yates_shuffle(indexed)
        new_opts = [t for t, _ in shuffled_idx]
        new_answer = next(i for t, i in shuffled_idx if t == correct_text)

        # 构建新对象字符串（保持原有格式风格）
        new_opts_str = ', '.join(f'"{o}"' for o in new_opts)
        new_obj = f'{{"type": "choice", "q": "{q_text}", "opts": [{new_opts_str}], "answer": {new_answer}, "hint": "{hint}"}}'

        # 替换原文
        start, end = m.start(), m.end()
        result = result[:start] + new_obj + result[end:]
        shuffled += 1

    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(result)

    print(f"  ✅ 完成！共{total}选择题，打乱{shuffled}题，跳过{skipped}题")

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    data_dir = os.path.join(script_dir, '..', 'data')
    files = ['4-2.js', '5-2.js', '6-2.js']

    for fname in files:
        fpath = os.path.join(data_dir, fname)
        if os.path.exists(fpath):
            process_file(fpath)
        else:
            print(f"文件不存在: {fpath}")

    print("\n🎉 全部处理完成！记得 git 提交推送。")

if __name__ == '__main__':
    main()
